docs/proposals/rescheduler.md

```diff
@@ -0,0 +1,130 @@
+
+Possible variations on this Baseline rescheduler are
+
+1. it can kill the Pod(s) whose space it wants **and also schedule the Pod that will take that space and reschedule the Pod(s) that were killed**, rather than just killing the Pod(s) whose space it wants and relying on the regular scheduler(s) to schedule the Pod that will take that space (and to reschedule the Pod(s) that were evicted)
+1. it can run continuously in the background to optimize general cluster layout instead of just trying to get a PENDING Pod to schedule
+1. it can try to move groups of Pods instead of using a one-at-a-time / greedy approach
+1. it can formulate multi-hop plans instead of single-hop
+
+A key design question for a Rescheduler is how much knowledge it needs about the scheduling policies used by the cluster's scheduler(s).
+* For the Baseline rescheduler, it needs to know the predicate functions used by the cluster's scheduler(s) else it can't know how to create a hole that the PENDING Pod will fit into, nor be sure that the evicted Pod(s) will be able to reschedule elsewhere.
+* If it is going to run continuously in the background to optimize cluster layout but is still only going to kill Pods, then it still needs to know the predicate functions for the reason mentioned above. In principle it doesn't need to know the priority functions; it could just randomly kill Pods and rely on the regular scheduler to put them back in better places. However, this is a rather inexact approach. Thus it is useful for the rescheduler to know the priority functions, or at least some subset of them, so it can be sure that an action it takes will actually improve the cluster layout.
+* If it is going to run continuously in the background to optimize cluster layout and is going to act as a scheduler rather than just killing Pods, then it needs to know the predicate functions and some compatible (but not necessarily identical) priority functions  One example of a case where "compatible but not identical" might be useful is if the main scheduler(s) has a very simple scheduling policy optimized for low scheduling latency, and the Rescheduler having a more sophisticated/optimal scheduling policy that requires more computation time. The main thing to avoid is for the scheduler(s) and rescheduler to have incompatible priority functions, as this will cause them to "fight" (though it still can't lead to an infinite loop, since the scheduler(s) only ever touches a Pod once).
+
+The vast majority of users probably only care about rescheduling for three scenarios:
+
+1. Redistribute Pods onto new nodes added by a cluster auto-scaler
+1. Move Pods around to get a PENDING Pod to schedule
+1. Move Pods around when CPU starvation is detected on a node
```

Comment:
I would start the document with use cases, as it drives motivation.  

+ the use case I mentioned above.